llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -2479,6 +2479,42 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
     if (match(Mag, m_FAbs(m_Value(X))) || match(Mag, m_FNeg(m_Value(X))))
       return replaceOperand(*II, 0, X);
 
+    Value *A, *B;
+    CmpInst::Predicate Pred;
+    const APFloat *TC, *FC;
+    if (!match(Sign, m_Select((m_And(m_FCmp(Pred, m_Value(A), m_PosZeroFP()),
+                                     m_Value(B))),
+                              m_APFloat(TC), m_APFloat(FC))))
+      return nullptr;
+
+    /*
+    Match select ?, TC, FC where the constants are equal but negated.
+    Check for these conditions:
+    olt/ule
+    copysign(Mag, B & (A < 0.0) ? -TC : TC) -> copysign(Mag, A) B->true, A<0.
+    copysign(Mag, B & (A < 0.0) ? -TC : TC) -> copysign(Mag, A) B->true, A>0.
+    copysign(Mag, B & (A < 0.0) ? -TC : TC) -> copysign(Mag, A) B->false, A>0.
+    copysign(Mag, B & (A < 0.0) ? -TC : TC) -> copysign(Mag, -A) B->false, A<0.
+    ogt/uge
+    copysign(Mag, B & !(A < 0.0) ? -TC : TC) -> copysign(Mag, -A) B->true, A<0.
+    copysign(Mag, B & !(A < 0.0) ? -TC : TC) -> copysign(Mag, -A) B->true, A>0.
+    copysign(Mag, B & !(A < 0.0) ? -TC : TC) -> copysign(Mag, A) B->false, A>0.
+    copysign(Mag, B & !(A < 0.0) ? -TC : TC) -> copysign(Mag, -A) B->false,A<0.
+    */
+
+    if (Pred == CmpInst::FCMP_OLT || Pred == CmpInst::FCMP_ULE) {
+      if (match(A, m_Negative()) && TC->isNegative())
```

Comment:
I don't think this m_Negative check works. Looking at PatternMatch, I think this only supports integers? Would it be worth add an AP variant of it?

Reply:
But that means this, and the m_ZeroInt checks below are dead code. You could open a separate PR to add an FP version of m_Negative. For FP values we already have m_AnyZeroFP, m_PosZeroFP and m_NegZeroFP